### PR TITLE
fixes #8133: migrate to setuptools_scm 6.x 

### LINKED ISSUE #8133: migrate to setuptools_scm >= 5
followup to #8101

part of the issue was that without the setuptools_scm 5.x release the per distribution version pretending is not availiable

### PR DESCRIPTION
**Pull request by RonnyPfannschmidt**
and use SETUPTOOLS_SCM_PRETEND_VERSION_FOR_PYTEST

fixes #8133

### CODE REVIEW

**Comment by RonnyPfannschmidt** [comment]
@bluetech i believe the missing install in the deploy command created the problem
do you by chance have around a link to the build job else i'll search for more details sometime in the afternoon

File: pyproject.toml [hunk 1]
```diff
@@ -2,7 +2,7 @@
 requires = [
   # sync with setup.py until we discard non-pep-517/518
   "setuptools>=42.0",
-  "setuptools-scm[toml]>=3.4",
+  "setuptools-scm[toml]>=5.0",
```
Comment (bluetech) [inline pyproject.toml:5]: Need this in the setup.cfg as well per comment above.

File: tox.ini [hunk 2]
```diff
@@ -84,9 +84,7 @@ commands =
 [testenv:regen]
 changedir = doc/en
 basepython = python3
-passenv = SETUPTOOLS_SCM_PRETEND_VERSION
-# TODO: When setuptools-scm 5.0.0 is released, use SETUPTOOLS_SCM_PRETEND_VERSION_FOR_PYTEST
-# and remove the next line.
+passenv = SETUPTOOLS_SCM_PRETEND_VERSION_FOR_PYTEST
 install_command=python -m pip --use-deprecated=legacy-resolver install {opts} {packages}
```
Comment (bluetech) [inline tox.ini:88]: The fix should enable the removal of the `--use-deprecated=legacy-resolver` here

**Review by bluetech — COMMENTED** [review]
Thanks @RonnyPfannschmidt. I didn't test SETUPTOOLS_SCM_PRETEND_VERSION_FOR_PYTEST on CI, only locally. I'll test this when I get a chance.

**Comment by bluetech** [comment]
Basically the test is `SETUPTOOLS_SCM_PRETEND_VERSION_FOR_PYTEST=10.20.30 tox -e regen` with a new pip and seeing that:

1. The regendoc install doesn't fail
2. The correct version appears in the doc

**Comment by RonnyPfannschmidt** [comment]
@bluetech thanks, i figured the issue - basically tox makes a sdist where the version pretending is not passed, which in turn confuses the rest

im now setting out to configure it correctly

**Comment by RonnyPfannschmidt** [comment]
@bluetech i found a bug in the setuptools_scm pyproject integration, bugfix upcoming

**Comment by RonnyPfannschmidt** [comment]
@gaborbernat is there potentially anything missing for tox to conisder it - even wit the bugfix i see the _FOR_PYTEST version being ignored while the normal one works

**Comment by RonnyPfannschmidt** [comment]
@gaborbernat - nm, after updating a externalsetuptools_scm twice it actually worked, im not sure how that came to be but i suspect i poisoned my user-site, sorry for the noise

**Review by nicoddemus — APPROVED** [review]
Should we backport this to `6.2.x`?

**Comment by RonnyPfannschmidt** [comment]
@bluetech please crosscheck if it works for you as well (to avoid a works on my box)

**Comment by RonnyPfannschmidt** [comment]
@nicoddemus good idea

**Comment by bluetech** [comment]
@RonnyPfannschmidt I checked out this branch and tried the following:

```
SETUPTOOLS_SCM_PRETEND_VERSION_FOR_PYTEST=10.20.30 tox -e regen
```

tox does say:

```
.package recreate: /home/ran/src/pytest/.tox/.package
.package installdeps: setuptools>=42.0, setuptools-scm[toml]>=5.0.1, wheel
```

but...

```
regen inst-nodeps: /home/ran/src/pytest/.tox/.tmp/package/1/pytest-6.3.0.dev38+g2753279b1.tar.gz
```

that is, the specified version doesn't take.

Strangely, even though `SETUPTOOLS_SCM_PRETEND_VERSION` (without the `FOR_PYTEST`) is no longer allowed in `passenv`, that one *does* work.

I tried to completely rm my `.tox` directory and verified I the this branch checked out, still happens. Though I might still be missing something.

**Comment by RonnyPfannschmidt** [comment]
Also update SETUPTOOLS_SCM in the surrounding env

**Comment by bluetech** [comment]
What do you mean by surrounding env?

**Comment by RonnyPfannschmidt** [comment]
@bluetech i appologize - the issue is observable for mae again as well

@gaborbernat - is there a easy way to debug what tox is doing here?

for some reason SETUPTOOLS_SCM_PRETEND_VERSION wont make it to the creation of the dist and i have no idea what exactly is wrong

**Comment by gaborbernat** [comment]
Did you passed in at passenv, and just call it with -vvv to see the stdout of the build.

**Comment by RonnyPfannschmidt** [comment]
i upgraded this to the most recent setuptools_scm which includes a bugfix
additionally i added extra details to the commit message

**Comment by RonnyPfannschmidt** [comment]
won't need a backport anymore

**Review by bluetech — APPROVED** [review]
I don't remember all of the details, but the changes LGTM.

**Comment by RonnyPfannschmidt** [comment]
@bluetech there was a combination issue between regendoc and pytest setuptools_scm pretend_version

a pretend version at install time for regendoc would set it up wrong,
and the just for pytest pretend version had a bug that was only just fixed in the latest setuptools_scm version

**Pull request merged**